[table] feat(Table2): new 'scroll' instance method

ccullotta opened the pull request:
#### Fixes #5915

Enables functionality which can address [this scroll issue](https://github.com/palantir/blueprint/issues/5915)

#### Checklist

- [n/a] Includes tests
- [ ] Update documentation

<!-- DO NOT enable CircleCI for your fork. Our build will run when you open this PR. -->

#### Changes proposed in this pull request:

- Add a new instance method `Table2.prototype.scroll()` which allows users to manually scroll the table by a given left / top offset in pixels
- Demonstrate usage of this new method in table-dev-app
- Provide an API to display a visual indicator that shows the user an affordance for auto-scrolling

#### Reviewers should focus on:
If this change makes sense as a "enable-new-feature" solution to the above issue, vs changing default table behavior. Also that the new "scroll" function behavior makes sense for general context.

<!-- Fill this out. -->

#### Screenshot

![2023-03-01 12 53 16](https://user-images.githubusercontent.com/723999/222221983-2bff91ad-f053-4bbd-bc14-8f6de28add6c.gif)

ccullotta committed:
added 'scroll' function to table2 properties to enable outside scroll modifications (v0)

ccullotta committed:
lint fix, also added some basic doc strings

adidahiya commented:
<h3>lint fix, also added some basic doc strings</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/0ae5b657-aea5-4239-b0cb-c2916ba04cea/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/0ae5b657-aea5-4239-b0cb-c2916ba04cea/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/0ae5b657-aea5-4239-b0cb-c2916ba04cea/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/0ae5b657-aea5-4239-b0cb-c2916ba04cea/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta committed:
ran prettier

adidahiya commented:
<h3>ran prettier</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/8fa925ab-b1d4-471f-9e68-44448333c9de/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/8fa925ab-b1d4-471f-9e68-44448333c9de/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/8fa925ab-b1d4-471f-9e68-44448333c9de/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/8fa925ab-b1d4-471f-9e68-44448333c9de/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta committed:
created basic demo

adidahiya commented:
<h3>created basic demo</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/9e132ec8-f3e4-4563-887d-1a4dc13df4c2/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/9e132ec8-f3e4-4563-887d-1a4dc13df4c2/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/9e132ec8-f3e4-4563-887d-1a4dc13df4c2/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/9e132ec8-f3e4-4563-887d-1a4dc13df4c2/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta committed:
making demo a little nicer

adidahiya commented:
<h3>making demo a little nicer</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/3fbbb564-da12-4a5f-ac16-d731dfab5d19/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/3fbbb564-da12-4a5f-ac16-d731dfab5d19/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/3fbbb564-da12-4a5f-ac16-d731dfab5d19/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/3fbbb564-da12-4a5f-ac16-d731dfab5d19/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta commented:
GIF of the solution:
![2023-02-24_17-56-23 (1)](https://user-images.githubusercontent.com/44371962/221315450-70a3617f-4684-4d5b-b742-a9d2542c2fb5.gif)

adidahiya committed:
Merge remote-tracking branch 'origin/develop' into cc/enable-scroll-control

adidahiya committed:
annotate type exports to avoid webpack warnings

adidahiya commented on packages/table/src/table2.tsx:
```diff
@@ -430,6 +430,29 @@ export class Table2 extends AbstractComponent2<Table2Props, TableState, TableSna
+     * - yOffset: The vertical scroll modification
+     *
+     */
+    public scroll(xOffset: number, yOffset: number) {
```
I think this should be `{ left: number, top: number }` to match the `Offset` type:

https://github.com/palantir/blueprint/blob/3c58026dbfdade2f573a81eb2aa4d85af573ed01/packages/popover2/src/contextMenu2Shared.ts#L21-L24

I will make a change later to reconcile them as the same type, but for now they should be assignable to each other for API consistency across Blueprint.

```suggestion
    public scroll(relativeOffset: { left: number; top: number }) {
```

adidahiya committed:
remove unnecessary copyright

adidahiya reviewed (changes requested):
In general, the idea of a `Table2.prototype.scroll` instance method sounds fine; it seems reasonable to add this next to the existing `scrollToRegion` instance method. But we'll need to do some work to make it more usable.

First off, there should be some visual indication that the user is triggering a scroll. One idea for this would be to add a linear gradient that fades from partially-transparent `$black` to fully-transparent `$black` at the top / bottom of the table's main quadrant. This could be supported with modifier classes on the `<Table2>` component, something like `Classes.TABLE_BODY_IS_SCROLLING_{TOP|LEFT|BOTTOM|TOP}`.

Secondly, the example needs to have improved scroll trigger behavior. When I move my mouse back towards the middle of the table (away from the 30% top or bottom regions), the scrolling should stop immediately. It shouldn't continue as an automatic animation.

![2023-02-27 15 23 50](https://user-images.githubusercontent.com/723999/221678929-2ae7a000-b1a3-4a24-ad5a-2d36faa4d9a1.gif)

For documentation, you should add to the "Instance methods" section of the table docs: https://github.com/palantir/blueprint/blob/3c58026dbfdade2f573a81eb2aa4d85af573ed01/packages/table/src/docs/table-api.md?plain=1#L17

adidahiya commented on packages/table/src/index.ts:
```diff
@@ -16,41 +16,41 @@
 /* eslint-disable deprecation/deprecation */
 
-export { Cell, CellProps, ICellProps, ICellRenderer, CellRenderer } from "./cell/cell";
+export { Cell, type CellProps, type ICellProps, type ICellRenderer, type CellRenderer } from "./cell/cell";
```
I was getting a bunch of unrelated webpack errors with these exports, and adding these `type` annotations seems to fix them:

<img width="1501" alt="Screenshot 2023-02-27 at 3 15 59 PM" src="https://user-images.githubusercontent.com/723999/221679730-f7d1592a-a977-4e32-9925-00909abaef9a.png">

adidahiya commented:
<h3>remove unnecessary copyright</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/c6ef2de5-4603-4e07-a11e-af50949b4873/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/c6ef2de5-4603-4e07-a11e-af50949b4873/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/c6ef2de5-4603-4e07-a11e-af50949b4873/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/c6ef2de5-4603-4e07-a11e-af50949b4873/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta committed:
fixed auto scrolling behavior

ccullotta committed:
Update packages/table/src/table2.tsx

Co-authored-by: Adi Dahiya <adahiya@palantir.com>

ccullotta committed:
merge in suggested parameter fix to enforce some consistency

ccullotta committed:
fixed small scroll bug

ccullotta committed:
add cursor scroll indicator

ccullotta committed:
got linear gradient working

ccullotta committed:
small fixes

ccullotta committed:
removed cruft and will-change edits

adidahiya commented:
<h3>small fixes</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/ab1a5375-24a1-41d0-b3e3-c88124550826/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/ab1a5375-24a1-41d0-b3e3-c88124550826/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/ab1a5375-24a1-41d0-b3e3-c88124550826/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/ab1a5375-24a1-41d0-b3e3-c88124550826/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

adidahiya commented:
<h3>removed cruft and will-change edits</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/0a936677-4aa9-4308-bf01-f7ca258f54f4/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/0a936677-4aa9-4308-bf01-f7ca258f54f4/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/0a936677-4aa9-4308-bf01-f7ca258f54f4/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/0a936677-4aa9-4308-bf01-f7ca258f54f4/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta committed:
reformatted setScrollOverlayBackground to be more customizable

ccullotta commented:
Updated to fix scrolling behavior to be more responsive. Added both a cursor, and a visual overlay affect to show the user is scrolling. Interested for feedback on setScrollOverlayBackground, I initially worked it as a setScrollIndicator (left, right, top,..) function, but decided that this seemed too rigid. Perhaps this is too custom though?

adidahiya commented:
<h3>reformatted setScrollOverlayBackground to be more customizable</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/a9d0fe95-85d4-48dd-bc49-652e9f11c2a9/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/a9d0fe95-85d4-48dd-bc49-652e9f11c2a9/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/a9d0fe95-85d4-48dd-bc49-652e9f11c2a9/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/a9d0fe95-85d4-48dd-bc49-652e9f11c2a9/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

adidahiya commented:
> Interested for feedback on setScrollOverlayBackground, I initially worked it as a setScrollIndicator (left, right, top,..) function, but decided that this seemed too rigid. Perhaps this is too custom though?

I think we should be opinionated about this and not provide customization until users ask for it and really need it.

Some initial thoughts:

1. when you've reached the end of the scroll bounds, we should probably stop showing the scroll indicator gradient.
2. can you make the gradient fade in and/or grow from the edges? it should transition in a bit more subtly, not pop in quickly like it does now

@CPerinet could we get some visual design help here? What do you think this feature should feel like?

<img width="507" alt="image" src="https://user-images.githubusercontent.com/723999/222221322-331aabbb-7e5e-48b8-b12a-ffa94ce38a40.png">

<img width="550" alt="image" src="https://user-images.githubusercontent.com/723999/222221341-962895fa-210d-49e8-aea4-63915709bc02.png">

ccullotta committed:
added fade in, and removed fade at top and bottom

ccullotta commented:
Added fade in, struggling a little with fade out though, will re-attack in the morning

adidahiya commented:
<h3>added fade in, and removed fade at top and bottom</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/0d4cd81b-797b-431b-ac5b-e0f068917933/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/0d4cd81b-797b-431b-ac5b-e0f068917933/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/0d4cd81b-797b-431b-ac5b-e0f068917933/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/0d4cd81b-797b-431b-ac5b-e0f068917933/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta committed:
added fade-out animation

adidahiya commented:
<h3>added fade-out animation</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/ecfd8156-b8e0-4d5b-9790-00ce687bd4c5/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/ecfd8156-b8e0-4d5b-9790-00ce687bd4c5/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/ecfd8156-b8e0-4d5b-9790-00ce687bd4c5/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/ecfd8156-b8e0-4d5b-9790-00ce687bd4c5/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta committed:
added back scroll at top behavior

adidahiya commented:
<h3>added back scroll at top behavior</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/a74447cc-bd70-4fa8-91ce-c3dedf99ccca/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/a74447cc-bd70-4fa8-91ce-c3dedf99ccca/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/a74447cc-bd70-4fa8-91ce-c3dedf99ccca/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/a74447cc-bd70-4fa8-91ce-c3dedf99ccca/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta committed:
fixed some docstrings and cruft

ccullotta committed:
fixed  lint issues

adidahiya commented:
<h3>fixed some docstrings and cruft</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/bf12a6a7-f4e5-401f-845d-e203b7388d33/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/bf12a6a7-f4e5-401f-845d-e203b7388d33/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/bf12a6a7-f4e5-401f-845d-e203b7388d33/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/bf12a6a7-f4e5-401f-845d-e203b7388d33/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta commented:
As part of the solution above, I created a new method: setScrollOverlayIndicator to manage the indicator direction from outside the table. Curious for feedback on that, I've been wondering if that should be merged into "scroll" as something like "programaticScroll(offset: {...}, direction: "UP" | "DOWN" ....)". The two methods are tightly coupled, but in the case the user stops scrolling midway, I think it would be strange to have a call like: programaticScroll({left: 0, top: 0}, direction "NONE")

adidahiya commented:
<h3>fixed  lint issues</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/26c442be-2499-4d21-b504-590e22038935/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/26c442be-2499-4d21-b504-590e22038935/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/26c442be-2499-4d21-b504-590e22038935/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/26c442be-2499-4d21-b504-590e22038935/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta commented:
![2023-03-02_07-08-39 (1)](https://user-images.githubusercontent.com/44371962/222424757-380b597d-8610-4584-9f11-70ba40bb1d87.gif)

ccullotta commented:
![2023-03-02_07-10-28 (1)](https://user-images.githubusercontent.com/44371962/222425200-21fe7d94-6f74-4320-9f13-3a2a56bcc871.gif)

ccullotta commented:
![2023-03-02_07-13-15 (1)](https://user-images.githubusercontent.com/44371962/222425754-a38d2ed5-2f8e-4e8c-91e4-7a19925a9437.gif)

ccullotta committed:
merged scroll and setScrolling indicator into setScrolling

ccullotta commented:
Update: I thought about it more, and I think coalescing the two functions, and renaming makes the most sense. Merged scroll and setScrollingIndicatorOverlay into setScrolling. The new function is simpler to use, and more concise code-wise than before.

adidahiya commented:
<h3>merged scroll and setScrolling indicator into setScrolling</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/6f8747c3-5841-4907-8a5c-2898b5d13eba/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/6f8747c3-5841-4907-8a5c-2898b5d13eba/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/6f8747c3-5841-4907-8a5c-2898b5d13eba/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/6f8747c3-5841-4907-8a5c-2898b5d13eba/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

adidahiya commented on packages/table/src/quadrants/tableQuadrantStack.tsx:
```diff
@@ -433,8 +435,16 @@ export class TableQuadrantStack extends AbstractComponent2<ITableQuadrantStackPr
             />
         ) : undefined;
 
+        const tableOverlay = () => {
```
nit: this should be a method on the class so that we don't generate a new method on every render:

```
private renderTableOverlay = () => {...}
```

ccullotta replied:
Absolutely

adidahiya commented on packages/table/src/table2.tsx:
```diff
@@ -430,6 +430,42 @@ export class Table2 extends AbstractComponent2<Table2Props, TableState, TableSna
+     * - scrollDirection the scrolling direction for which we should generate the scroll indicator
+     *   gradient.
+     */
+    public setScrolling(
```
nice, this is cleaner as one method instead of two.

can we call this `scrollByOffset()`?

also it seems like we can infer the scroll direction from the relative offset, right? so you shouldn't have to provide the second parameter.

and for "NONE", we can accept a `null` value which cancels scrolling

```ts
public scrollByOffset(relativeOffset: { left: number; top: number } | null);
```

ccullotta replied:
Implemented!

adidahiya commented on packages/table/src/table2.tsx:
```diff
@@ -1098,6 +1156,49 @@ export class Table2 extends AbstractComponent2<Table2Props, TableState, TableSna
+     */
+    private renderScrollIndicatorOverlay = (scrollBarWidth: number, columnHeaderHeight: number) => {
+        const { scrollDirection } = this.state;
+        const handleStyle = (direction: string | undefined, compare: string) => {
```
nit: it's not an event handler, so don't use `handle...` terminology:

```suggestion
        const getStyle = (direction: string | undefined, compare: string) => {
```

adidahiya commented on packages/table/src/tableState.ts:
```diff
@@ -94,6 +94,8 @@ export interface TableState {
 
     childrenArray: Array<React.ReactElement<ColumnProps>>;
+
+    scrollDirection?: "TOP" | "BOTTOM" | "LEFT" | "RIGHT" | "NONE";
```
let's save this type as a const enum somewhere in the table package so that it's not duplicated in multiple places:

```ts
export enum ScrollDirection {
  TOP = "top",
  BOTTOM = "bottom",
  // ...
}
```

ccullotta replied:
makes sense!

adidahiya commented on packages/table/src/table2.tsx:
```diff
@@ -430,6 +430,42 @@ export class Table2 extends AbstractComponent2<Table2Props, TableState, TableSna
     }
 
+    /**
+     * Can adjust the scroll offset the table by the provided x and y values. Can also set a
+     * linear gradient overlay effect to the table to indicate scrolling with the scrolling
+     * direction provided.
```
```suggestion
     * Scrolls the table by a specified number of offset pixels in either the horizontal or vertical dimension.
```

adidahiya commented on packages/table/src/table2.tsx:
```diff
@@ -430,6 +430,42 @@ export class Table2 extends AbstractComponent2<Table2Props, TableState, TableSna
+     * linear gradient overlay effect to the table to indicate scrolling with the scrolling
+     * direction provided.
+     *
+     * - relativeOffset: The horizontal and vertical scroll modification to be applied.
```
use JSDoc to describe parameters

```suggestion
     * @param relativeOffset - how much to scroll the table body in pixels relative to the current scroll offset
```

adidahiya reviewed (changes requested):
nice, this is looking pretty close!

I would suggest adding this to the main mutableTable example, with a toggle to enable the behavior (titled something like "Demo programmatic scrolling API" in the sidebar)

some small polish items, if possible (might be ok to address as follow-ups):

1. sometimes the scroll indicator sticks around after you mouse off and reach the end of the table, like at the end of this gif:

![2023-03-02 11 47 17](https://user-images.githubusercontent.com/723999/222500721-174f36cc-3772-462f-835c-66ec8a96264f.gif)

2. sometimes it keeps scrolling after I've moused off quickly, like this:

![2023-03-02 11 47 42](https://user-images.githubusercontent.com/723999/222500871-e9efd243-e359-42dc-a12d-1edaa381e2f7.gif)

ccullotta committed:
Update packages/table/src/table2.tsx

Co-authored-by: Adi Dahiya <adahiya@palantir.com>

ccullotta committed:
Update packages/table/src/table2.tsx

Co-authored-by: Adi Dahiya <adahiya@palantir.com>

ccullotta committed:
Update packages/table/src/table2.tsx

Co-authored-by: Adi Dahiya <adahiya@palantir.com>

ccullotta committed:
updated names, small refactors, improved scroll indicator to go away if not scrolling.

ccullotta committed:
merged into table demo

ccullotta commented:
was able to fix the first issue, second one is still eluding me, but hoping we can fix that as a flup? Also merged my example into the existing with the switch being "Demo programmatic scrolling API"

ccullotta committed:
adding enum that was accidentally ommitted

adidahiya commented:
<h3>adding enum that was accidentally ommitted</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/2181b04f-0f79-471a-9297-ad15308dbf5b/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/2181b04f-0f79-471a-9297-ad15308dbf5b/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/2181b04f-0f79-471a-9297-ad15308dbf5b/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/2181b04f-0f79-471a-9297-ad15308dbf5b/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

adidahiya commented on packages/table-dev-app/src/mutableTable.tsx:
```diff
@@ -265,6 +266,9 @@ export interface IMutableTableState {
     showRowHeadersLoading?: boolean;
     showTableInteractionBar?: boolean;
     showZebraStriping?: boolean;
+    scrollDirection?: "UP" | "DOWN";
+    animationRequestId?: number;
+    previousTime?: number;
```
nit: these don't need to live in component state, because they don't affect the `render()` code path. they can just be private class properties instead.

ccullotta replied:
mind blown! thanks for the info!

adidahiya commented on packages/table-dev-app/src/mutableTable.tsx:
```diff
@@ -349,7 +357,8 @@ export class MutableTable extends React.Component<{}, IMutableTableState> {
                         <div className={layoutBoundary ? "layout-boundary" : "layout-passthrough-fill"}>
-                            {this.renderTable()}
+                            {this.renderTable()};
+                            {/* empty string above to get mutable table to render a smaller demo table */}
```
this comment is no longer relevant

ccullotta replied:
ack! - removing cruft

adidahiya commented on packages/table-dev-app/src/mutableTable.tsx:
```diff
@@ -390,48 +399,109 @@ export class MutableTable extends React.Component<{}, IMutableTableState> {
+                    display: "flex",
+                    flexDirection: "row",
+                    gap: 20,
+                    height: 300,
```
this changes the default height for the mutable table, which I'd like to avoid doing in this PR. I guess you're reducing the height so that you can  more easily demo the feature.

can you change this code so that the demo only uses this reduced height when `enableScrollingApi` is enabled? or, even better, perhaps: automatically set `inline: true` when `enableScrollingApi` is true (and make the "Inline" switch disabled when you do this)

<img width="309" alt="image" src="https://user-images.githubusercontent.com/723999/222756624-e5a06ef2-0fcb-4354-9ae5-83ea43938f19.png">

ccullotta replied:
removed the height styling entirely and applied the listener to the surrounding layout boundary div. I thought the height was needed to actually be able to scroll in the table, but it turned out to be an inadvertent bug, which actually was causing the inline table to be unscrollable, both the the api and to users. Fixed, and now both the inline table and the default work as expected, with and without demoing the programmatic scrolling

adidahiya commented on packages/table-dev-app/src/mutableTable.tsx:
```diff
@@ -390,48 +399,109 @@ export class MutableTable extends React.Component<{}, IMutableTableState> {
-                cellRendererDependencies={[this.state.cellContent]}
+            <div
+                ref={this.refHandlers.tableWrapperRef}
+                style={{
```
please write these styles in CSS rather than inline styles. there's are some other classes with custom styling in `table-dev-app/src/index.scss`

ccullotta replied:
thanks! moving those now..

adidahiya reviewed (changes requested):
almost there!

yes it's fine to fix that second issue as a FLUP

ccullotta committed:
adding function to base table, and adding documentation to blueprint instance methods api

ccullotta committed:
removing cruft

ccullotta committed:
removing inline style cruft in mutable table

ccullotta committed:
removing styles we didnt need, making the reduced height only apply when demoing the scrolling api

ccullotta committed:
fixed bug causing inline switch to not work as intended

ccullotta committed:
removing cruft

ccullotta committed:
removing more cruft

adidahiya commented:
<h3>removing more cruft</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/b66fe9e3-82b2-4f86-b8f5-363ac5138c9c/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/b66fe9e3-82b2-4f86-b8f5-363ac5138c9c/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/b66fe9e3-82b2-4f86-b8f5-363ac5138c9c/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/b66fe9e3-82b2-4f86-b8f5-363ac5138c9c/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ccullotta committed:
making certain state properties private class properties

ccullotta commented:
Copied over the instance method from table 2 into table as well, apparently it was required for the documentation to port into the instance method docs, and I assume that both should have it anyhow. Also added the documentation. There is a little funkiness with the formatting though, @adidahiya do you have any dots?

<img width="1077" alt="image" src="https://user-images.githubusercontent.com/44371962/222785409-931277f9-b2a0-4dbe-9e2c-e8cff9d6120f.png">

ccullotta commented:
The above is capitalized correctly in the commits I made ftr, just forgot to recompile

adidahiya commented:
<h3>making certain state properties private class properties</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/29c0f619-f0a8-4a9b-a540-96ff183683d8/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/29c0f619-f0a8-4a9b-a540-96ff183683d8/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/29c0f619-f0a8-4a9b-a540-96ff183683d8/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/29c0f619-f0a8-4a9b-a540-96ff183683d8/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

adidahiya commented on packages/table/src/docs/table-api.md:
```diff
@@ -33,6 +33,8 @@ type ICellMapper<T> = (rowIndex: number, columnIndex: number) => T;
 
 @method Table.scrollToRegion
 
+@method Table.scrollByOffset
```
I think you can avoid adding this method to Table and only use Table2:

```suggestion
@method Table2.scrollByOffset
```

ccullotta replied:
*facepalm* that makes sense

ccullotta replied:
for some reason I thought it had to be on table 1

adidahiya commented on packages/table/src/table.tsx:
```diff
@@ -402,6 +403,52 @@ export class Table extends AbstractComponent2<TableProps, TableState, TableSnaps
         this.quadrantStackInstance.scrollToPosition(correctedScrollLeft, correctedScrollTop);
     }
 
+    /**
```
I'd prefer to not add code to Table in this PR, just limit the changes to Table2

ccullotta replied:
ack that checks out

adidahiya reviewed (commented):
looks good in table-dev-app preview!

re: typedef in this bit of the docs... the docs-theme is a bit buggy when it comes to this... I'll try to fix it in a follow-up PR

<img width="342" alt="image" src="https://user-images.githubusercontent.com/723999/223197621-b65eecd2-81a6-4e73-b518-f4c8d98e7a84.png">

ccullotta committed:
remove table update

ccullotta committed:
use table2 for documentation string

adidahiya commented:
<h3>use table2 for documentation string</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/1dec4edb-8884-4f4a-9c6a-f82adcfedd7d/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/1dec4edb-8884-4f4a-9c6a-f82adcfedd7d/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/1dec4edb-8884-4f4a-9c6a-f82adcfedd7d/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/1dec4edb-8884-4f4a-9c6a-f82adcfedd7d/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>
